modin/core/storage_formats/pandas/aggregations.py

```diff
@@ -0,0 +1,590 @@
+            A tuple holding the Map (at the first position) and the Reduce (at the second position) kernels
+            computing correlation/covariance matrix.
+        """
+        if method == cls.Method.COV:
+            raise NotImplementedError("Computing covariance is not yet implemented.")
+
+        return _CorrCovKernels.map, lambda df: _CorrCovKernels.reduce(
+            df, min_periods, method
+        )
+
+
+class _CorrCovKernels:
+    """Holds kernel functions computing correlation/covariance matrices in a MapReduce manner."""
+
+    @classmethod
+    def map(cls, df: pandas.DataFrame) -> pandas.DataFrame:
+        """
+        Perform the Map phase to compute the corr/cov matrix.
+
+        In this kernel we compute all the required components to compute
+        the correlation matrix at the reduce phase, the required components are:
+            1. Matrix holding sums of pairwise multiplications between all columns
+               defined as ``M[col1, col2] = sum(col1[i] * col2[i] for i in range(col_len))``
+            2. Sum for each column (special case if there are NaN values)
+            3. Sum of squares for each column (special case if there are NaN values)
+            4. Number of values in each column (special case if there are NaN values)
+
+        Parameters
+        ----------
+        df : pandas.DataFrame
+            Partition to compute the aggregations for.
+
+        Returns
+        -------
+        pandas.DataFrame
+            A MultiIndex columned DataFrame holding the described aggregation results for this
+            specifix partition under the following keys: ``["mul", "sum", "pow2_sum", "count"]``
+        """
+        df = df.select_dtypes(include="number")
+        # It's more convenient to use a NumPy array here as it appears to perform
+        # much faster in for-loops which this kernel function has plenty of
+        raw_df = df.values.T
+        nan_mask = np.isnan(raw_df)
+        has_nans = nan_mask.sum() != 0
+
+        if has_nans:
+            if not raw_df.flags.writeable:
+                # making a copy if the buffer is read-only
+                raw_df = raw_df.copy()
+            # Replacing all NaNs with zeros so we can use much
+            # faster `np.sum()` instead of slow `np.nansum()`
+            np.putmask(raw_df, nan_mask, values=0)
+
+        cols = df.columns
+        # Here we compute a sum of pairwise multiplications between all columns
+        # result:
+        #   col1: [sum(col1 * col2), sum(col1 * col3), ... sum(col1 * colN)]
+        #   col2: [sum(col2 * col3), sum(col2 * col4), ... sum(col2 * colN)]
+        #   ...
+        sum_of_pairwise_mul = pandas.DataFrame(
+            np.dot(raw_df, raw_df.T), index=cols, columns=cols, copy=False
+        )
+
+        if has_nans:
+            sums, sums_of_squares, count = cls._compute_nan_aggs(raw_df, cols, nan_mask)
+        else:
+            sums, sums_of_squares, count = cls._compute_non_nan_aggs(df)
```

Comment:
Why do we operate a pandas df in _compute_non_nan_aggs, but a np array in _compute_nan_aggs? Why can't we use np arrays in both functions?

Reply:
that's because it's more convenient to use either pandas dataframe or a numpy array depending on the circumstances :)

- In `compute_non_nan_aggs` it's more convenient & faster [to use pandas API](https://github.com/dchigarev/modin/blob/9b5f220f57675781739d11052aa64608902e5ca8/modin/core/storage_formats/pandas/aggregations.py#L218-L222) in order to compute the aggregations
- When in the `compute_nan_aggs` we [perform complex transformations](https://github.com/dchigarev/modin/blob/9b5f220f57675781739d11052aa64608902e5ca8/modin/core/storage_formats/pandas/aggregations.py#L268-L288) that are done much more easily & faster in numpy